File: approximating-eulers-number-correctly/ApproximateEInterval.java
/* 
 * Approximating Euler's number correctly (Java)
 * 
 * Copyright (c) 2014 Nayuki Minase
 * All rights reserved. Contact Nayuki for licensing.
 * http://nayuki.eigenstate.org/page/approximating-eulers-number-correctly
 */

import java.math.BigInteger;


public class ApproximateEInterval {
	
	// Runs a demo that prints out some numbers
	public static void main(String[] args) {
		// Print e rounded to n decimal places, for n from 0 to 60
		for (int i = 0; i <= 60; i++)
			System.out.println(computeEulersNumber(i));
		System.out.println();
		
		// Compute 1 to 100000 decimal places (exponentially increasing) and print timing
		int prev = 0;
		for (int i = 0; i <= 100; i++) {
			int digits = (int)Math.round(Math.pow(10, i / 20.0));
			if (digits == prev)
				continue;
			prev = digits;
			
			long startTime = System.nanoTime();
			computeEulersNumber(digits);
			long time = System.nanoTime() - startTime;
			System.out.printf("%6d  %7.3f s%n", digits, time / 1000000000.0);
		}
	}
	
	
	// For example: computeEulersNumber(4) = "2.7183"
	private static String computeEulersNumber(int accuracy) {
		if (accuracy < 0)
			throw new IllegalArgumentException();
		
		for (int extraPrecision = 7; ; extraPrecision += 2) {
			String result = computeEulersNumber(accuracy, extraPrecision);
			if (result != null)
				return result;
		}
	}
	
	
	private static String computeEulersNumber(int accuracy, int extraPrecision) {
		final BigInteger fullScaler  = BigInteger.TEN.pow(accuracy + extraPrecision);
		final BigInteger extraScaler = BigInteger.TEN.pow(extraPrecision);
		BigInteger sumLow  = BigInteger.ZERO;
		BigInteger sumHigh = BigInteger.ZERO;
		BigInteger termLow  = fullScaler;
		BigInteger termHigh = fullScaler;
		
		for (int i = 0; termLow.signum() != 0; i++) {
			sumLow  = sumLow .add(termLow );
			sumHigh = sumHigh.add(termHigh);
			
			if (i >= 1 && termHigh.compareTo(extraScaler) < 0) {
				BigInteger sumUpperBound = sumHigh.add(termHigh);
				BigInteger temp = divideAndRound(sumLow, extraScaler);
				if (divideAndRound(sumUpperBound, extraScaler).equals(temp)) {
					// Note: The number of terms used is i+1
					String s = temp.toString();
					return s.substring(0, s.length() - accuracy) + "." + s.substring(s.length() - accuracy);
				}
			}
			
			BigInteger j = BigInteger.valueOf(i + 1);
			termLow  = termLow .divide(j);
			termHigh = termHigh.divide(j).add(BigInteger.ONE);
		}
		return null;
	}
	
	
	// Any rounding mode works correctly with computeEulersNumber().
	// Round-half-to-even is implemented here, but truncation, flooring, etc. are acceptable too.
	private static BigInteger divideAndRound(BigInteger num, BigInteger div) {
		BigInteger[] quotRem = num.divideAndRemainder(div);
		BigInteger result = quotRem[0];
		int cmp = quotRem[1].shiftLeft(1).compareTo(div);
		if (cmp > 0 || cmp == 0 && result.testBit(0))
			result = result.add(BigInteger.ONE);
		return result;
	}
	
}

File: approximating-eulers-number-correctly/approximate-e-interval.py
# 
# Approximating Euler's number correctly (Python)
# 
# Copyright (c) 2025 Project Nayuki
# All rights reserved. Contact Nayuki for licensing.
# https://www.nayuki.io/page/approximating-eulers-number-correctly
# 

import time
from typing import Optional


# Runs a demo that prints out some numbers
def main() -> None:
	# Print e rounded to n decimal places, for n from 0 to 60
	for i in range(61):
		print(compute_eulers_number(i))
	print()
	
	# Compute 1 to 100000 decimal places (exponentially increasing) and print timing
	prev: int = 0
	for i in range(101):
		digits: int = int(round(10 ** (i / 20)))
		if digits == prev:
			continue
		prev = digits
		
		start_time: float = time.time()
		compute_eulers_number(digits)
		elapsed_time: float = time.time() - start_time
		print(f"{digits:6}  {elapsed_time:7.3f} s")


# For example: compute_eulers_number(4) = "2.7183"
def compute_eulers_number(accuracy: int) -> str:
	if accuracy < 0:
		raise ValueError()
	
	extra_precision: int = 7
	while True:
		result: Optional[str] = compute_eulers_number_internal(accuracy, extra_precision)
		if result is not None:
			return result
		extra_precision += 2


def compute_eulers_number_internal(accuracy: int, extra_precision: int) -> Optional[str]:
	full_scaler: int = 10 ** (accuracy + extra_precision)
	extra_scaler: int = 10 ** extra_precision
	
	sum_low : int = 0
	sum_high: int = 0
	term_low : int = full_scaler
	term_high: int = full_scaler
	i: int = 0
	while term_low > 0:
		sum_low  += term_low
		sum_high += term_high
		
		if (i >= 1) and (term_high < extra_scaler):
			sum_upper_bound: int = sum_high + term_high
			temp: int = divide_and_round(sum_low, extra_scaler)
			if divide_and_round(sum_upper_bound, extra_scaler) == temp:
				# Note: The number of terms used is i+1
				s: str = str(temp)
				return s[ : len(s) - accuracy] + "." + s[len(s) - accuracy : ]
		
		i += 1
		term_low  = term_low  // i
		term_high = term_high // i + 1
	return None


# Any rounding mode works correctly with compute_eulers_number_internal().
# Round-half-to-even is implemented here, but truncation, flooring, etc. are acceptable too.
def divide_and_round(num: int, div: int) -> int:
	quot, rem = divmod(num, div)
	if (rem * 2 > div) or ((rem * 2 == div) and (quot & 1 == 1)):
		quot += 1
	return quot


if __name__ == "__main__":
	main()
